include/nanobind/eigen/sparse.h

```diff
@@ -0,0 +1,118 @@
+    using StorageIndexCaster = make_caster<StorageIndexTensor>;
+
+    NB_TYPE_CASTER(T, const_name<row_major>("scipy.sparse.csr_matrix[",
+                                            "scipy.sparse.csc_matrix[")
+                   + make_caster<Scalar>::Name + const_name("]"));
+
+    ScalarCaster scalar_caster;
+    StorageIndexCaster storageindex_caster;
+
+    bool from_python(handle src, uint8_t flags, cleanup_list *cleanup) noexcept {
+        object obj = borrow(src);
+        try {
+            object matrix_type = module_::import_("scipy.sparse").attr(row_major ? "csr_matrix" : "csc_matrix");
+            if (!obj.type().is(matrix_type))
+                obj = matrix_type(obj);
+        } catch (const python_error &) {
+            return false;
+        }
+
+        if (object data_o = obj.attr("data"); !scalar_caster.from_python(data_o, flags, cleanup))
+            return false;
+        ScalarTensor& values = scalar_caster.value;
+
+        if (object indices_o = obj.attr("indices"); !storageindex_caster.from_python(indices_o, flags, cleanup))
+            return false;
+        StorageIndexTensor& inner_indices = storageindex_caster.value;
+
+        if (object indptr_o = obj.attr("indptr"); !scalar_caster.from_python(indptr_o, flags, cleanup))
+            return false;
+        StorageIndexTensor& outer_indices = storageindex_caster.value;
+
+        object shape_o = obj.attr("shape"), nnz_o = obj.attr("nnz");
+        Index rows, cols, nnz;
+        try {
+            if (len(shape_o) != 2)
+                return false;
+            rows = cast<Index>(shape_o[0]);
+            cols = cast<Index>(shape_o[1]);
+            nnz = cast<Index>(nnz_o);
+        } catch (const python_error &) {
+            return false;
+        }
+
+        value = SparseMap(rows, cols, nnz, outer_indices.data(), inner_indices.data(), values.data());
+
+        return true;
+    }
+
+    static handle from_cpp(T &&v, rv_policy policy, cleanup_list *cleanup) noexcept {
+        if (policy == rv_policy::automatic ||
+            policy == rv_policy::automatic_reference)
+            policy = rv_policy::move;
+
+        return from_cpp((const T &) v, policy, cleanup);
+    }
+
+    static handle from_cpp(const T &v, rv_policy policy, cleanup_list *cleanup) noexcept {
+	if (!v.isCompressed()) {
+	    PyErr_SetString(PyExc_ValueError,
+			    "nanobind: unable to return an Eigen sparse matrix that is not in a compressed format. "
+			    "Please call `.makeCompressed()` before returning the value on the C++ end.");
+	    return handle();
+	}
+        T & src = const_cast<T &>(v);
+
+        object matrix_type = module_::import_("scipy.sparse").attr(row_major ? "csr_matrix" : "csc_matrix");
+
+        const size_t data_shape[] = { (size_t)v.nonZeros() };
+        const size_t outer_indices_shape[] = { (size_t)((row_major ? v.rows() : v.cols()) + 1) };
+        ScalarTensor data(src.valuePtr(), 1, data_shape);
```

Comment:
`const_cast` is still needed because `nb::tensor` has no constructor for `const void *`. Is there a deeper reason for that?

Reply:
we currently don't have a way of making read-only tensors, though that is a feature I would like to support in the future.